### ISSUE TITLE
Custom output writer

### ISSUE THREAD

**Issue by sn00pster** [bblanchon/ArduinoJson#1088]
I am using the library on an application that does not use Arduino, therefore to output to a custom “output” i am limited to either using a buffer or using ostream.

I cannot use ostream as including it takes my flash usage up by an amount to the point that my application doesn’t fit into flash anymore.

There should be a lightweight class, much like the String stream writer which can be subclassed and used as the output.

i’m currently having to patch the source of the library to add my own class to achieve this.

In situations where you don’t have Arduino Print class or ostream then a lightweight alternative would be a great solution.

Ideally after serialization is complete Arduino json would call a function in the class to finalize, this would allow and underlying data that hasnt been output to be output, in the case of writing to a block device directly.

**Comment by bblanchon** [comment]
Hi @sn00pster,

I think I can do that by adding an overload of `serializeJson()`.
Then you'll just have to implement the two `write()` methods.

Best Regards,
Benoit

**Comment by sn00pster** [comment]
> Hi @sn00pster,
> 
> I think I can do that by adding an overload of `serializeJson()`.
> Then you'll just have to implement the two `write()` methods.
> 
> Best

That would be great, honestly a super lightweight serialization would be a life saver.

I assume the same could be done for deserialization?

I'm aware that this probably seems like a bit of an odd request given the excellent mechanisms already provided, but when using the library in an severely constrained system this lightweight mechanism would be a life saver.  I'm sure many others will find it useful as well!

**Comment by bblanchon** [comment]
This feature is available on branch 6.x.
Your custom writer class must look like that:

```c++
struct CustomWriter {
  size_t write(uint8_t c);
  size_t write(const uint8_t *s, size_t n);
};
```

See [CustomWriter.cpp](https://github.com/bblanchon/ArduinoJson/blob/498a2e4c1e5294468071bc4814940ef939284a9e/extras/tests/JsonSerializer/CustomWriter.cpp) for a complete example.

**Comment by sn00pster** [comment]
Wow, that was fast!  Does this also work for MsgPack?

Thanks

**Comment by bblanchon** [comment]
Yes, it works with MessagePack as well.
You're welcome ;-)

**Comment by bblanchon** [comment]
This feature is available in [ArduinoJson version 6.13.0](https://github.com/bblanchon/ArduinoJson/releases/tag/v6.13.0)
